### DISCUSSION TITLE
Gray background while converting pdf

### DISCUSSION THREAD

**Discussion by Zzzenk** [ImageMagick/ImageMagick#4797] [Help] [answered]
Hi there!

I try to convert .pdf to .png and output files have gray background instead of white.
Mac OS 12.2 / PHP 8.1 / Imagick 3.7.0

I try this code on Windows - all OK. How to fix it on Mac?

**Code:**
```
$imagick = new Imagick();
$imagick->setResolution(150, 150);
$imagick->readImage($row);
$imagick->setImageFormat("png");
$imagick->writeImages($uploaddir . uniqid($filename) . ".png", true);
```
**Here's a result:**
![HSK_3 19-361fa6a70a87f1-0](https://user-images.githubusercontent.com/2630902/152240443-2e1f6b39-9c2a-4fbe-95f5-bfca60a54b0c.png)

**UPD1:**
Also tried with `Spatie\PdfToImage\Pdf` - same result.

**UPD2:**
PDF:
[HSK 3.19-3.pdf](https://github.com/ImageMagick/ImageMagick/files/7990683/HSK.3.19-3.pdf)

**Comment by fmw42** [comment]
Are you sure the background is not transparent and you are seeing it on a display device that colors it gray?

Post your original PDF (you can zip it if necessary), so others can test it. 

Do you have Ghostscript installed and if so, what  version.  ImageMagick requires Ghostscript to read PDF files.

**Comment by Zzzenk** [comment]
I attach output image, so, no, background are not transparent.

Attached pdf into my first post.

Ghostscript 9.55.0 (2021-09-27).

**Comment by fmw42** [comment]
Your PDF has transparency.  In ImageMagick command line, I would do the following to get the first page:

`convert HSK.3.19-3.pdf[0] -background white -alpha background -alpha off x.png`

This will work for all pages of the PDF.

![x](https://user-images.githubusercontent.com/13458218/152257383-ce7ce774-99e3-4a3f-a392-ad7f97869d52.png)

Sorry, I do not use Imagick or any other API.

Alternately, on a page by page basis only, you could do

`convert HSK.3.19-3.pdf[0] -background white -flatten x.png`

or

`convert -background white HSK.3.19-3.pdf[0] x.png`

**Try setting the background color to white before reading the PDF in your Imagick.**

**Reply by Zzzenk** [reply]
Very strange. I tried you command-line command, but result the same - gray background.

I also tried `$imagick->setBackgroundColor('white');` or `$imagick->setBackgroundColor('#ffffff');` before readImage - same result.

**Reply by fmw42** [reply]
Sorry, I do not use Imagick. Check your Imagick and ImageMagick version and perhaps upgrade.

**Reply by fmw42** [reply]
Try the command line using PHP exec(). If that works, then it is your Imagick or the version of ImageMagick used by Imagick or the Ghostscript linked to that version of ImageMagick.

**Reply by Zzzenk** [reply]
PHP `exec()` - same result...

**Reply by fmw42** [reply]
What was your exact PHP exec command?

What do you get from

```
exec("gs --version")

also

exec("convert -version")

or

exec("magick -version")

also

exec("which -a convert")

or

exec("which -a magick")

```
Do you have more than one version of Ghostscript installed?

**Reply by Zzzenk** [reply]
My exec command:
`exec("convert /System/Volumes/Data/Users/zenk/Sites/sinotati/HSK_3.19-3.pdf -background white -alpha background -alpha off 12341234.png", $qwe, $qwer);`

Your commands result is followong:
```
string(6) "9.55.0"
string(19) "Compiler: gcc (4.2)"
string(19) "Compiler: gcc (4.2)"
string(22) "/usr/local/bin/convert"
string(21) "/usr/local/bin/magick"
```

I installed Ghostscript one time.

**Comment by fredlcore** [comment]
I'm having the same problem on Mac OS Monterey. The attached simple Word file produces the attached PDF when using this command line:
`gm convert -density 300 Dokument1.pdf test.png`
[Dokument1.pdf](https://github.com/ImageMagick/ImageMagick/files/8004152/Dokument1.pdf)
![test](https://user-images.githubusercontent.com/24660955/152554914-2ac2f1e6-918a-48d2-b2cd-293e2e3b2f9f.png)

It was working until a few days ago and I didn't change anything in my setup as far as I remember. Any solution yet?

**Reply by fredlcore** [reply]
Strange, when using imagemagick's convert instead of graphicsmagic `gm convert` and then adding the above mentioned options:
`-background white -alpha background -alpha off`
then it works. Unfortunately, gm convert does not support `-alpha`, so I had to install imagemagick. If anyone has a workaround for gm (and/or an explanation why this problem suddenly arises), I'd be happy to know!

**Reply by fmw42** [reply]
gm convert is graphicsmagick, which is a different product not supported by ImageMagick. So its code may not be the same as ImageMagick and it probably does not support -background white -alpha background -alpha off. Please ask on a GraphiccsMagick forum about it.

**Reply by fredlcore** [reply]
Oh, sorry, thanks!

**Reply by fredlcore** [reply]
BTW: I'm wondering if the underlying issue is a different one because with both @Zzzenk and me, the issue occurs on MacOS and (at least for me) only since a couple of days (not sure if that's the same for @Zzzenk, but at least the post is just a couple of days old). Maybe it's a Mac system library of some sort that was affected which leads to the same result in both GraphicsMagick and ImageMagick? Just a thought, I'm happy with the workaround that was provided here.

**Reply by fmw42** [reply]
I am running on Mac OSX Sierra with no issues. What version of Ghostscript do you have installed?

**Reply by fredlcore** [reply]
9.55.0 on Mac OS 12.1 Monterey, that's the latest that `brew` gives me...

**Comment by fmw42** [comment]
When you do

```
exec("convert -version")

or

exec("magick -version")

```
You are not getting back correct information. Something is wrong if all you get is compiler versions. 

> string(19) "Compiler: gcc (4.2)"
> string(19) "Compiler: gcc (4.2)"
> 

So there is something wrong with your ImageMagick installs.

**Reply by Zzzenk** [reply]
Any ideas how to fix it?
I have istalled imagick and ghostscript via brew and I didn't find any other manuals.

**Reply by fmw42** [reply]
Can you test directly from the command line with 

```
convert -version

or

magick -version

```

What version of ImageMagick did you install from Homebrew?

**Reply by fmw42** [reply]
Did you install Imagick and PHP? Perhaps it is just a PHP issue?

**Reply by Zzzenk** [reply]
```
Version: ImageMagick 7.1.0-22 Q16-HDRI x86_64 2022-01-29 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(5.0) 
Delegates (built-in): bzlib fontconfig freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib
Compiler: gcc (4.2)
```

Yes, I have installed PHP 8.1

**Reply by fmw42** [reply]
The ImageMagick version looks fine.

Can you test directly from the command line with

```

magick image.pdf[0] -background white -alpha background -alpha off x_0.png


```

Is it white or gray?

**Reply by Zzzenk** [reply]
Gray

**Reply by fmw42** [reply]
What about (again simply from the command line)

`magick -background white image.pdf[0] x_0.png
`

**Reply by fredlcore** [reply]
`brew`did not seem to install `magick`, only `convert`. And the latter results in a grey background with these commands.

**Reply by fmw42** [reply]
Then the commands should use convert and not magick.  Which were you using?  If Brew installed convert that would be ImageMagick 6 and you could not have 7.1.0-22 from that install.  Perhaps you have two versions of ImageMagick, one version 6 and one version 7, and one has an older version of Ghostscript than you think.

**Reply by fredlcore** [reply]
Sorry, I just replied to the e-mail and didn't notice that this was a question to @Zzzenk. He has 7.1.0-22.

**Comment by fmw42** [comment]
The only other workaround that I can suggest is to process each page separately, then put them back together as a new PDF.

`convert image.pdf[x] -background white -flatten result_x.png
`
Where x is 0, 1, 2, 3 ... n one page at a time. Then put all the pngs back as a new PDF.

Try 

`convert image.pdf[0] -background white -flatten result_0.png
`
Is that gray or white?

**Reply by fredlcore** [reply]
For me, `-background white -flatten` is sufficient, no need for `-alpha background -alpha off`, but without `-background white` the image background will be gray...

**Reply by fmw42** [reply]
What was your full command line?

**Reply by fredlcore** [reply]
`convert -background white -flatten Dokument1.pdf "PNG32:result_0.png"`
but even just
`convert -flatten Dokument1.pdf  "PNG32:result_0.png"`
works

**Reply by fmw42** [reply]
That is because the default -background color is white.

**Reply by fredlcore** [reply]
Ah, ok.

**Reply by fredlcore** [reply]
Unfortunately, I just encountered another PDF "type" that I need to convert where the barcode is the only thing in the PDF that keeps the gray background (which makes it unreadable of course), the rest of the document becomes white. In the original PDF, everything is/looks white background. Do you have any idea what kind of other setting could affect this behaviour? The barcode in that PDF is not an image because I can select the individual bars in the PDF with the mouse, if that is of any help.

**Reply by fmw42** [reply]
Post an example.  It could be that the barcode is an image embedded in the PDF.

**Reply by fredlcore** [reply]
Since this is a shipping label which includes  both personal information (addresses) as well as the potentially reusable barcode, could I send it to you via e-mail?

**Reply by fmw42** [reply]
One of the IM developers will need to give you the correct email address.

**Comment by fmw42** [comment]
The image HSK.3.19-3.pdf has a completely opaque alpha channel. When I do

`magick HSK.3.19-3.pdf[0] x.png`

I get a white background color.

![x](https://user-images.githubusercontent.com/13458218/152586622-fd277708-1cba-4a1e-b743-1ead8b2958a4.png)

I do not know why you get a gray one. Perhaps it is Ghostscript

What do you get from 

`magick HSK.3.19-3.pdf[0] -alpha off x.png`

**Reply by Zzzenk** [reply]
Also gray background(

OK, thx for your help, I'll create an issue.

**Reply by fmw42** [reply]
What do you get if you process the image directly with Ghostscript?

**Comment by fmw42** [comment, marked as answer]
Perhaps it is the Ghostscript arguments set in your delegates.xml file.  What do you have for PS encode PDF? Mine is

  `<delegate decode="ps" encode="pdf" mode="bi" command="&quot;gs&quot; -sstdout=%%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 &quot;-sDEVICE=pdfwrite&quot; &quot;-sOutputFile=%o&quot; &quot;-f%i&quot;"/>
`

**Reply by Zzzenk** [reply]
Seems like the same...
usr/local/etc/ImageMagick-7/delegates.xml
`  <delegate decode="ps" encode="pdf" mode="bi" command="&quot;gs&quot; -sstdout=%%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 &quot;-sDEVICE=pdfwrite&quot; &quot;-sOutputFile=%o&quot; &quot;-f%i&quot;"/>`

**Reply by fmw42** [reply]
What are you using to view the file?

**Reply by Zzzenk** [reply]
To view delegates.xml? PhpStorm. Why?

**Reply by fmw42** [reply]
Does it look the same (gray or white) if you do

`magick display x.png`

after converting from pdf to png?

**Reply by Zzzenk** [reply]
Background is not transparent, if you are asking about it.

And `magick display` is not working:
`display: delegate library support not built-in '' (X11) @ error/display.c/DisplayImageCommand/1898.`

**Reply by fmw42** [reply]
What is your OS?  If on Mac, install XQuartz to be able to use the display.

**Reply by Zzzenk** [reply]
magick display - still gray

**Reply by fmw42** [reply]
I cannot explain it. It is white for me on my Mac OSX Sierra with ImageMagick 6.9.12-37 and 7.1.0-22. I displayed in Mac Preview, ImageMagick display and GraphicConverter. All show white.

**Reply by Zzzenk** [reply]
So the only difference is Sierra and Monterey

**Reply by fmw42** [reply]
Presumably. But I installed my delegates from MacPorts and then installed ImageMagick from source. You did yours from Homebrew. But I do not know your Ghostscript environment. I am on GS 9.54. I think you said you had 9.55. So perhaps there is a difference.

**Reply by fredlcore** [reply]
It would explain why the problem occured only a couple of days ago. @Zzzenk, did you also encounter this only recently or did you just start using it recently and had the problem from the outset?
However, I didn't do a `brew upgrade gs`, I assume that was the version when I set up my computer in December, but I can't say for sure. Is there a way to downgrade to 9.54 via `brew`? Then I could test this.

**Reply by fmw42** [reply]
Sorry, I do not use Brew, so cannot say. But I would assume you could do that. Check the Brew installation notes or search Google about it.

**Reply by fmw42** [reply]
@fredlcore.  Are you also having this issue and do you have GS 9.55 also?

**Reply by fredlcore** [reply]
Depends what issue you mean, I have the issue with both GraphicsMagick and ImageMagick that all of a sudden I get the grey background when background is apparently transparent. Changing the command line parameters as mentioned above fixes this issue for ImageMagick's convert, haven't found the appropriate options for gm yet. But since it occurs in both programs and for both @Zzzenk and me, I assume that it has to do with gs and or Monterey...

**Reply by fredlcore** [reply]
I just checked `brew info gs` and it tells me that 9.55 was installed mid-December (about the time when I set up my new laptop). So a version change is rather unlikely then...

**Reply by Zzzenk** [reply]
> Presumably. But I installed my delegates from MacPorts and then installed ImageMagick from source. You did yours from Homebrew. But I do not know your Ghostscript environment. I am on GS 9.54. I think you said you had 9.55. So perhaps there is a difference.

@fmw42 I solve the problem:
1) Uninstall imagick & ghostscript
2) Install from MacPorts

Not it's converting correctly. Seems like something is missed in homebrew package...

Thank you for your awesome help!

**Reply by fredlcore** [reply]
Thanks for the update, can you post the versions of imagick and ghostscript you are using now compared to the ones used by brew?

**Reply by Zzzenk** [reply]
> Thanks for the update, can you post the versions of imagick and ghostscript you are using now compared to the ones used by brew?

https://ports.macports.org/port/ImageMagick/
ImageMagick
v 6.9.11-60

**Reply by Zzzenk** [reply]
But now I'm wondering how to add this version of imagick to my php?
@fmw42

**Reply by fmw42** [reply]
FYI. That is a rather older version of ImageMagick 6.

**Reply by fmw42** [reply]
See https://www.php.net/manual/en/book.imagick.php

**Comment by fmw42** [comment]
@Zzzenk

Any difference if you add -strip to your command?

**Reply by fmw42** [reply]
I suggest you either post your issue to a Ghostscript forum or to StackOverflow. At the latter, it seems that KenS is the expert. Perhaps he can clarify if it is 9.55 vs 9.54

**Reply by fmw42** [reply]
You could also post a bug report for Ghostscript at https://bugs.ghostscript.com

**Comment by fmw42** [comment]
```
Uninstall imagick & ghostscript
Install from MacPorts
Now it's converting correctly. Seems like something is missed in homebrew package...

```

You probably got IM 7 from Homebrew and MacPorts only supports IM 6.  However, I tried both releases and got a white image.  

What version of GS did MacPorts install?  What is now on your system?

**Reply by Zzzenk** [reply]
GS 9.55.0

**Reply by fmw42** [reply]
Then it is either Homebrew or the version of ImageMagick with your OS. It really does not make sense to me. I even tried the same version of ImageMagick that you had before 7.1.0-22 and 7.1.0-20 and it is always white for me.

Perhaps GS 9.55 from Homebrew has an issue but is OK from MacPorts.

**Comment by jbogdani** [comment]
Hello all,
I'm having the ssame issue, on Monterey running ImageMagick 7.1.0-25, installed via HomeBrew. Tried several combiniations of `-background white`, `-alpha background`, `-alpha off`, `-flatten` with no success....

**Reply by fmw42** [reply]
From my comments above, I suspect something in Homebrew. Perhaps their copy of Ghostscript.

**Comment by Zzzenk** [comment]
It's me again. I have new MacBook Pro (16', 2021), and after casual installation all works fine...
`brew install pkg-config imagemagick`
`pecl install imagick`

I guess, it was black magic.

**Comment by sesu-bio** [comment]
I had a simular issue with GhostScript on Linux. The root cause was a [bug in Little-CMS 2.13](https://github.com/mm2/Little-CMS/commit/fdbfb7694f9d7048d53674b79ddfc38068bfdaf7) which is now fixed in 2.13.1.
